File: pkg/dev_compiler/test/codegen_expected/lib/typed_data/float32x4_clamp_test.js
dart_library.library('lib/typed_data/float32x4_clamp_test', null, /* Imports */[
  'dart_sdk',
  'expect'
], function load__float32x4_clamp_test(exports, dart_sdk, expect) {
  'use strict';
  const core = dart_sdk.core;
  const typed_data = dart_sdk.typed_data;
  const dart = dart_sdk.dart;
  const dartx = dart_sdk.dartx;
  const expect$ = expect.expect;
  const float32x4_clamp_test = Object.create(null);
  let VoidTovoid = () => (VoidTovoid = dart.constFn(dart.definiteFunctionType(dart.void, [])))();
  let VoidTodynamic = () => (VoidTodynamic = dart.constFn(dart.definiteFunctionType(dart.dynamic, [])))();
  float32x4_clamp_test.testClampLowerGreaterThanUpper = function() {
    let l = typed_data.Float32x4.new(1.0, 1.0, 1.0, 1.0);
    let u = typed_data.Float32x4.new(-1.0, -1.0, -1.0, -1.0);
    let z = typed_data.Float32x4.zero();
    let a = z.clamp(l, u);
    expect$.Expect.equals(a.x, 1.0);
    expect$.Expect.equals(a.y, 1.0);
    expect$.Expect.equals(a.z, 1.0);
    expect$.Expect.equals(a.w, 1.0);
  };
  dart.fn(float32x4_clamp_test.testClampLowerGreaterThanUpper, VoidTovoid());
  float32x4_clamp_test.testClamp = function() {
    let l = typed_data.Float32x4.new(-1.0, -1.0, -1.0, -1.0);
    let u = typed_data.Float32x4.new(1.0, 1.0, 1.0, 1.0);
    let z = typed_data.Float32x4.zero();
    let a = z.clamp(l, u);
    expect$.Expect.equals(a.x, 0.0);
    expect$.Expect.equals(a.y, 0.0);
    expect$.Expect.equals(a.z, 0.0);
    expect$.Expect.equals(a.w, 0.0);
  };
  dart.fn(float32x4_clamp_test.testClamp, VoidTovoid());
  float32x4_clamp_test.main = function() {
    for (let i = 0; i < 2000; i++) {
      float32x4_clamp_test.testClampLowerGreaterThanUpper();
      float32x4_clamp_test.testClamp();
    }
  };
  dart.fn(float32x4_clamp_test.main, VoidTodynamic());
  // Exports:
  exports.float32x4_clamp_test = float32x4_clamp_test;
});

File: tests/lib/typed_data/float32x4_clamp_test.dart
// Copyright (c) 2013, the Dart project authors.  Please see the AUTHORS file
// for details. All rights reserved. Use of this source code is governed by a
// BSD-style license that can be found in the LICENSE file.
// VMOptions=--optimization-counter-threshold=10 --no-background-compilation

// Library tag to be able to run in html test framework.
library float32x4_clamp_test;

import 'dart:typed_data';
import 'package:expect/expect.dart';

void testClampLowerGreaterThanUpper() {
  Float32x4 l = new Float32x4(1.0, 1.0, 1.0, 1.0);
  Float32x4 u = new Float32x4(-1.0, -1.0, -1.0, -1.0);
  Float32x4 z = new Float32x4.zero();
  Float32x4 a = z.clamp(l, u);
  Expect.equals(a.x, 1.0);
  Expect.equals(a.y, 1.0);
  Expect.equals(a.z, 1.0);
  Expect.equals(a.w, 1.0);
}

void testClamp() {
  Float32x4 l = new Float32x4(-1.0, -1.0, -1.0, -1.0);
  Float32x4 u = new Float32x4(1.0, 1.0, 1.0, 1.0);
  Float32x4 z = new Float32x4.zero();
  Float32x4 a = z.clamp(l, u);
  Expect.equals(a.x, 0.0);
  Expect.equals(a.y, 0.0);
  Expect.equals(a.z, 0.0);
  Expect.equals(a.w, 0.0);
}

Float32x4 negativeZeroClamp() {
  final negZero = -Float32x4.zero();
  return negZero.clamp(negZero, Float32x4.zero());
}

Float32x4 zeroClamp() {
  final negOne = -Float32x4(1.0, 1.0, 1.0, 1.0);
  return Float32x4.zero().clamp(negOne, -Float32x4.zero());
}

// Regression test for https://github.com/dart-lang/sdk/issues/40426.
void testNegativeZeroClamp(Float32x4 unopt) {
  final res = negativeZeroClamp();
  Expect.equals(res.x.compareTo(unopt.x), 0);
  Expect.equals(res.y.compareTo(unopt.y), 0);
  Expect.equals(res.z.compareTo(unopt.z), 0);
  Expect.equals(res.w.compareTo(unopt.w), 0);
}

// Regression test for https://github.com/dart-lang/sdk/issues/40426.
void testZeroClamp(Float32x4 unopt) {
  final res = zeroClamp();
  Expect.equals(res.x.compareTo(unopt.x), 0);
  Expect.equals(res.y.compareTo(unopt.y), 0);
  Expect.equals(res.z.compareTo(unopt.z), 0);
  Expect.equals(res.w.compareTo(unopt.w), 0);
}

main() {
  final unoptNegZeroClamp = negativeZeroClamp();
  final unoptZeroClamp = zeroClamp();
  for (int i = 0; i < 2000; i++) {
    testClampLowerGreaterThanUpper();
    testClamp();
    testNegativeZeroClamp(unoptNegZeroClamp);
    testZeroClamp(unoptZeroClamp);
  }
}
